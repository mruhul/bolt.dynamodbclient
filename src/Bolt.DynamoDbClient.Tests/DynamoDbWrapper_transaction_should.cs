using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using Bolt.DynamoDbClient.Tests.Builders;
using Bolt.DynamoDbClient.Tests.TestHelpers;

namespace Bolt.DynamoDbClient.Tests;

public class DynamoDbWrapper_transaction_should
{
    private IDynamoDbWrapper sut;
    private IAmazonDynamoDB fake;

    public DynamoDbWrapper_transaction_should()
    {
        fake = Substitute.For<IAmazonDynamoDB>();
        sut = DynamoDbWrapperBuilder.Build(fake);
    }

    [Fact]
    public async Task call_dynamo_with_correct_request()
    {
        var transaction = sut.Transaction();

        transaction.Create(new SampleRecord
        {
            Pk = "sample-record#1",
            Sk = "sample-record#details",
            Name = "testing"
        });

        transaction.Upsert(new SampleRecord
        {
            Pk = "sample-record#2",
            Sk = "sample-record#details",
            Name = "testing 2"
        });

        transaction.Update(new SampleRecord
        {
            Pk = "sample-other-record#1",
            Sk = "sample-other-record#details",
            Name = "testing other record update 1"
        });

        transaction.Increment<SampleRecordIncrement>(
            "inc-pk-1",
            "inc-sk-1",
            [
                new("TotalCommits", 1),
                new("TotalLikes", -1),
            ]);

        TransactWriteItemsRequest? gotRequest = null;
        fake.TransactWriteItemsAsync(Arg.Any<TransactWriteItemsRequest>(), Arg.Any<CancellationToken>())
            .Returns(new TransactWriteItemsResponse())
            .AndDoes(c => { gotRequest = c.Arg<TransactWriteItemsRequest>(); });


        await transaction.Execute(CancellationToken.None);

        gotRequest.ShouldMatchApproved();
    }

    public record SampleRecord : SampleRecordBase
    {
        public string Name { get; set; }
    }

    public record SampleOtherRecord : SampleRecordBase
    {
        public string Age { get; set; }
        public string Token { get; set; }
    }

    public record SampleRecordIncrement : SampleRecordBase
    {
        public int TotalCommits { get; set; }
        public int TotalLikes { get; set; }
    }

    [DynamoDbTable("sample-store")]
    public record SampleRecordBase
    {
        [DynamoDbColumn("PK")] public string Pk { get; set; } = string.Empty;
        [DynamoDbColumn("SK")] public string Sk { get; set; } = string.Empty;
    }
}