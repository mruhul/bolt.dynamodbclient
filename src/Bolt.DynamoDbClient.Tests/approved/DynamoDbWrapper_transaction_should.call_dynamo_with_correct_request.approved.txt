{
  "transactItems": [
    {
      "put": {
        "conditionExpression": "attribute_not_exists(#pk) and attribute_not_exists(#sk)",
        "expressionAttributeNames": {
          "#pk": "PK",
          "#sk": "SK"
        },
        "expressionAttributeValues": {},
        "item": {
          "Name": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "testing",
            "ss": [],
            "isSSSet": false
          },
          "PK": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "sample-record#1",
            "ss": [],
            "isSSSet": false
          },
          "SK": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "sample-record#details",
            "ss": [],
            "isSSSet": false
          }
        },
        "tableName": "sample-store"
      }
    },
    {
      "put": {
        "expressionAttributeNames": {},
        "expressionAttributeValues": {},
        "item": {
          "Name": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "testing 2",
            "ss": [],
            "isSSSet": false
          },
          "PK": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "sample-record#2",
            "ss": [],
            "isSSSet": false
          },
          "SK": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "sample-record#details",
            "ss": [],
            "isSSSet": false
          }
        },
        "tableName": "sample-store"
      }
    },
    {
      "update": {
        "expressionAttributeNames": {
          "#Name": "Name"
        },
        "expressionAttributeValues": {
          ":Name": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "testing other record update 1",
            "ss": [],
            "isSSSet": false
          }
        },
        "key": {
          "PK": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "sample-other-record#1",
            "ss": [],
            "isSSSet": false
          },
          "SK": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "sample-other-record#details",
            "ss": [],
            "isSSSet": false
          }
        },
        "tableName": "sample-store",
        "updateExpression": "SET #Name = :Name"
      }
    },
    {
      "update": {
        "expressionAttributeNames": {
          "#prop0": "TotalCommits",
          "#prop1": "TotalLikes"
        },
        "expressionAttributeValues": {
          ":start": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "n": "0",
            "ns": [],
            "isNSSet": false,
            "null": false,
            "ss": [],
            "isSSSet": false
          },
          ":incr0": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "n": "1",
            "ns": [],
            "isNSSet": false,
            "null": false,
            "ss": [],
            "isSSSet": false
          },
          ":incr1": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "n": "-1",
            "ns": [],
            "isNSSet": false,
            "null": false,
            "ss": [],
            "isSSSet": false
          }
        },
        "key": {
          "PK": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "inc-pk-1",
            "ss": [],
            "isSSSet": false
          },
          "SK": {
            "bool": false,
            "isBOOLSet": false,
            "bs": [],
            "isBSSet": false,
            "l": [],
            "isLSet": false,
            "m": {},
            "isMSet": false,
            "ns": [],
            "isNSSet": false,
            "null": false,
            "s": "inc-sk-1",
            "ss": [],
            "isSSSet": false
          }
        },
        "tableName": "sample-store",
        "updateExpression": "SET #prop0 = if_not_exists(#prop0, :start) \u002B :incr0, #prop1 = if_not_exists(#prop1, :start) \u002B :incr1"
      }
    }
  ]
}